### PR TITLE
Add APM instrumentation

### PR DESCRIPTION
**Pull request by jsoriano**
In https://github.com/elastic/package-registry/issues/670 we are planning to introduce some refactors, in part motivated by performance issues. Introduce some instrumentation to evaluate what parts of the service require more performance tunning. And in the future to be able to compare the impact of the applied changes.

With this instrumentation we can collect traces such as this one:
Captura de pantalla de 2021-07-20 19-36-32

Checklist:
- [x] Fix `/categories`, tests failing.
- [x] Add docs.

### CODE REVIEW

File: util/packages.go [hunk 1]
```diff
@@ -25,20 +27,26 @@ type Packages []Package
 // The list is stored in memory and on the second request directly served from memory.
 // This assumes changes to packages only happen on restart (unless development mode is enabled).
 // Caching the packages request many file reads every time this method is called.
-func GetPackages(packagesBasePaths []string) (Packages, error) {
+func GetPackages(ctx context.Context, packagesBasePaths []string) (Packages, error) {
+	span, ctx := apm.StartSpan(ctx, "GetPackages", "custom")
+	defer span.End()
```
Comment (axw) [inline util/packages.go:32]: Given that the function just returns cached results if not already loaded, maybe it's not worth creating a span for this? It'll mostly be extremely fast, and when it's not then all the time will be attributed to GetPackagesFromFilesystem

File: categories.go [hunk 2]
```diff
@@ -26,129 +28,163 @@ type Category struct {
...
 
-			// Skip experimental packages if flag is not specified
-			if p.Release == util.ReleaseExperimental && !experimental {
-				continue
-			}
+	return filter, nil
+}
+
+func (filter categoriesFilter) Filter(ctx context.Context, packages util.Packages) map[string]util.Package {
+	span, _ := apm.StartSpan(ctx, "FilterPackages", "custom")
```
Comment (axw) [inline categories.go:98]: I would suggest changing all the span types from "custom" to "app". This should work better with the "time spent by span type" chart in APM:

image

Transaction time is countered under "app", and if you add spans with type "app" then they will add to this. i.e. it's best practice (probably not written down anywhere) to use "app" for spans related to application code.

**Review by axw — APPROVED** [review]
Looks good from the Go Agent usage angle, just a couple of minor comments.

File: categories.go (hunk 2 above)
Comment (mtojek) [inline categories.go:98]: nit: I assume it's `error` swallowed. Is it safe here?

**Review by mtojek — APPROVED** [review]
LGTM as long as CI is happy. Do we need to modify Dockerfiles in Package Storage too?

File: util/packages.go (hunk 1 above)
Comment (jsoriano) [reply util/packages.go]: Good point. Currently the packages are cached in memory, and only reloaded on startup, but this is something that can change in the near future. There may be a need to "reload" packages, what could affect this method, or they may be cached in some shared external store.
Said that, we can add the span when this method does more things.

File: categories.go (hunk 2 above)
Comment (jsoriano) [reply categories.go]: Changed :+1: 

Btw, when initializing the service ([here](https://github.com/elastic/package-registry/pull/702/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R88)), I am using "backend.init" as transaction type, is this ok?
Are there some guidelines or list of recommended values for transaction and span types?

**Comment by jsoriano** [comment]
> Do we need to modify Dockerfiles in Package Storage too?

Yes, next step would be to see where we send the data, and configure the registries used in production.

File: categories.go (hunk 2 above)
Comment (axw) [reply categories.go]: That is okay, and I don't think it particularly matters. HTTP requests will be given the transaction type "request". I'm not aware of any guidance for transaction types for this kind of program initialisation.

**Pull request merged**